tx-pool/src/component/relative.rs

```diff
@@ -0,0 +1,80 @@
+use crate::component::container::SortedTxMap;
+use ckb_types::packed::ProposalShortId;
+use std::collections::HashSet;
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) enum Relation {
+    Ancestor,
+    Descendant,
+}
+
+/// This crate is used for recursively searching in-pool ancestors or descendants
+pub(crate) trait RelativeTraversal {
+    fn get_relative_ids(
+        &self,
+        id: &ProposalShortId,
+        relation: Relation,
+    ) -> HashSet<ProposalShortId>;
+
+    fn get_parents(&self, id: &ProposalShortId) -> &HashSet<ProposalShortId>;
+
+    fn get_children(&self, id: &ProposalShortId) -> &HashSet<ProposalShortId>;
+
+    fn get_ancestors(&self, id: &ProposalShortId) -> HashSet<ProposalShortId> {
+        self.get_relative_ids(id, Relation::Ancestor)
+    }
+
+    fn get_descendants(&self, id: &ProposalShortId) -> HashSet<ProposalShortId> {
+        self.get_relative_ids(id, Relation::Descendant)
+    }
+}
+
+impl RelativeTraversal for SortedTxMap {
+    fn get_parents(&self, id: &ProposalShortId) -> &HashSet<ProposalShortId> {
+        &self
```

Comment:
return Option instead of panic. then use `result.map(ToOwned::to_owned).unwrap_or_default` outside.

Reply:
If an entry is in the tx_pool, then `get_parents` will be ok;
If an entry is not in the pool, then `get_parents` get panicked.

It is the duty for the outside caller to make sure that 
The outside caller should make ensure that the given entry is already in tx_pool before calling `RelativeTraversal's methods. I think the assert style is helpful for enforcing the consistency.

Reply:
Because `RelativeTraversal` is a public trait, and `get_parents ` is a public function. panic from a public function is not usually the rust idiomatic way, as you see in the std the more common style is `map.get(i).unwrap()`, the invoker control when to panic or not. This function looks don't like it'll panic, this is a pitfall for the invoker.